src/java.base/share/classes/java/lang/invoke/MethodHandle.java

```diff
@@ -855,34 +854,119 @@ public Object invokeWithArguments(java.util.List<?> arguments) throws Throwable
+        }
+        return at;
+    }
+
     /** Override this to change asType behavior. */
     /*non-public*/
     MethodHandle asTypeUncached(MethodType newType) {
-        if (!type.isConvertibleTo(newType))
-            throw new WrongMethodTypeException("cannot convert "+this+" to "+newType);
-        return asTypeCache = MethodHandleImpl.makePairwiseConvert(this, newType, true);
+        if (!type.isConvertibleTo(newType)) {
+            throw new WrongMethodTypeException("cannot convert " + this + " to " + newType);
+        }
+        return MethodHandleImpl.makePairwiseConvert(this, newType, true);
+    }
+
+    /**
+     * Returns true if {@code newType} does not depend on any class loader other than current method handle already does.
+     * May conservatively return false in order to be efficient.
+     */
+    private boolean isSafeToCache(MethodType newType) {
+        ClassLoader loader = getApproximateCommonClassLoader(type);
+        return keepsAlive(newType, loader);
+    }
+
+    /* Returns true when {@code loader} keeps {@code mt} either directly or indirectly through the loader delegation chain. */
+    private static boolean keepsAlive(MethodType mt, ClassLoader loader) {
+        for (Class<?> ptype : mt.ptypes()) {
+            if (!keepsAlive(ptype, loader)) {
+                return false;
+            }
+        }
+        return keepsAlive(mt.rtype(), loader);
+    }
+
+    /* Returns true when {@code loader} keeps {@code cls} either directly or indirectly through the loader delegation chain. */
+    private static boolean keepsAlive(Class<?> cls, ClassLoader loader) {
+        return keepsAlive(cls.getClassLoader(), loader);
+    }
+
+    /**
+     * Tries to find the most specific {@code ClassLoader} which keeps all the classes mentioned in {@code mt} alive.
+     * In the worst case, returns a {@code ClassLoader} which relates to some of the classes mentioned in {@code mt}.
+     */
+    private static ClassLoader getApproximateCommonClassLoader(MethodType mt) {
+        ClassLoader loader = ClassLoaders.appClassLoader();
+        if (keepsAlive(mt.rtype().getClassLoader(), loader)) {
+            loader = mt.rtype().getClassLoader();
+        }
+        for (Class<?> ptype : mt.ptypes()) {
+            ClassLoader ploader = ptype.getClassLoader();
+            if (loader != ploader && keepsAlive(loader, ploader)) {
+                loader = ploader;
+            } else {
+                // Either loader is a descendant of ploader or loaders are unrelated. Ignore both cases.
+                // When loaders are not related, just pick one and proceed. It reduces the precision of keepsAlive, but
+                // doesn't compromise correctness.
+            }
+        }
+        return loader;
+    }
+
+    /* Determine whether {@code descendant} keeps {@code ancestor} alive through the loader delegation chain. */
+    private static boolean keepsAlive(ClassLoader ancestor, ClassLoader descendant) {
```

Comment:
This method is not exactly doing `isAncestor` check.  It returns true if `ancestor` is a builtin loader even it's not an ancestor of `descendent`.   I agree that it would be helpful if the method is named by what it is.

Maybe naming it `isAncestor` but move the `isSystemLoader(ancestor)` check out to  the caller.  Just a thought.